File: backend-java/src/main/java/com/wealthtech/crm/modules/customer/service/ClientExcelService.java
package com.wealthtech.crm.modules.customer.service;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.time.LocalDate;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import org.apache.poi.ss.usermodel.Cell;
import org.apache.poi.ss.usermodel.CellStyle;
import org.apache.poi.ss.usermodel.CellType;
import org.apache.poi.ss.usermodel.DataFormatter;
import org.apache.poi.ss.usermodel.DateUtil;
import org.apache.poi.ss.usermodel.FillPatternType;
import org.apache.poi.ss.usermodel.Font;
import org.apache.poi.ss.usermodel.HorizontalAlignment;
import org.apache.poi.ss.usermodel.IndexedColors;
import org.apache.poi.ss.usermodel.Row;
import org.apache.poi.ss.usermodel.Sheet;
import org.apache.poi.ss.usermodel.Workbook;
import org.apache.poi.ss.usermodel.WorkbookFactory;
import org.apache.poi.xssf.usermodel.XSSFWorkbook;
import org.springframework.stereotype.Service;

import com.wealthtech.crm.modules.customer.dto.CreateClientRequest;
import com.wealthtech.crm.modules.customer.enums.Gender;

import lombok.extern.slf4j.Slf4j;

@Slf4j
@Service
public class ClientExcelService {

    private static final String[] HEADERS = {
        "First Name", "Last Name", "Email", "Phone", "PAN",
        "Date of Birth (YYYY-MM-DD)", "Gender (MALE/FEMALE/OTHER)",
        "Address Line", "City", "State", "Pincode", "Country"
    };

    private final DataFormatter dataFormatter = new DataFormatter();

    /**
     * Generates a styled Excel (.xlsx) template with headers and 1 sample row.
     */
    public byte[] generateClientBulkTemplate() {
        try (Workbook workbook = new XSSFWorkbook();
             ByteArrayOutputStream out = new ByteArrayOutputStream()) {

            Sheet sheet = workbook.createSheet("Clients Template");

            // Header Style
            CellStyle headerStyle = workbook.createCellStyle();
            Font headerFont = workbook.createFont();
            headerFont.setBold(true);
            headerFont.setColor(IndexedColors.WHITE.getIndex());
            headerStyle.setFont(headerFont);
            headerStyle.setFillForegroundColor(IndexedColors.DARK_BLUE.getIndex());
            headerStyle.setFillPattern(FillPatternType.SOLID_FOREGROUND);
            headerStyle.setAlignment(HorizontalAlignment.CENTER);

            // Create Header Row
            Row headerRow = sheet.createRow(0);
            for (int i = 0; i < HEADERS.length; i++) {
                Cell cell = headerRow.createCell(i);
                cell.setCellValue(HEADERS[i]);
                cell.setCellStyle(headerStyle);
            }

            // Create Sample Row
            Row sampleRow = sheet.createRow(1);
            sampleRow.createCell(0).setCellValue("Aarav");
            sampleRow.createCell(1).setCellValue("Sharma");
            sampleRow.createCell(2).setCellValue("aarav.sharma@example.com");
            sampleRow.createCell(3).setCellValue("9876543210");
            sampleRow.createCell(4).setCellValue("ABCDE1234F");
            sampleRow.createCell(5).setCellValue("1992-05-15");
            sampleRow.createCell(6).setCellValue("MALE");
            sampleRow.createCell(7).setCellValue("Flat 402, Highline Towers");
            sampleRow.createCell(8).setCellValue("Bengaluru");
            sampleRow.createCell(9).setCellValue("Karnataka");
            sampleRow.createCell(10).setCellValue("560001");
            sampleRow.createCell(11).setCellValue("India");

            // Auto-size columns
            for (int i = 0; i < HEADERS.length; i++) {
                sheet.autoSizeColumn(i);
            }

            workbook.write(out);
            return out.toByteArray();
        } catch (Exception e) {
            log.error("Failed to generate bulk client upload template", e);
            throw new RuntimeException("Could not generate Excel template", e);
        }
    }

    /**
     * Parses uploaded .xlsx rows into List<CreateClientRequest>.
     */
    public List<CreateClientRequest> parseClientExcel(InputStream inputStream) {
        List<CreateClientRequest> clients = new ArrayList<>();

        try (Workbook workbook = WorkbookFactory.create(inputStream)) {
            Sheet sheet = workbook.getSheetAt(0);

            for (int i = 1; i <= sheet.getLastRowNum(); i++) {
                Row row = sheet.getRow(i);
                if (row == null || isRowEmpty(row)) {
                    continue;
                }

                String firstName = getCellValueAsString(row.getCell(0));
                String lastName = getCellValueAsString(row.getCell(1));
                String email = getCellValueAsString(row.getCell(2));
                String phone = getCellValueAsString(row.getCell(3));
                String pan = getCellValueAsString(row.getCell(4));
                LocalDate dob = parseDate(row.getCell(5));
                Gender gender = parseGender(getCellValueAsString(row.getCell(6)));
                String addressLine = getCellValueAsString(row.getCell(7));
                String city = getCellValueAsString(row.getCell(8));
                String state = getCellValueAsString(row.getCell(9));
                String pincode = getCellValueAsString(row.getCell(10));
                String country = getCellValueAsString(row.getCell(11));

                if (firstName == null || email == null) {
                    continue; // Skip invalid rows
                }

                CreateClientRequest clientRequest = new CreateClientRequest(
                        firstName,
                        lastName != null ? lastName : "",
                        email,
                        phone,
                        pan,
                        dob != null ? dob : LocalDate.of(1990, 1, 1),
                        gender,
                        null, // RM will be assigned as the current user
                        addressLine,
                        city,
                        state,
                        pincode,
                        (country != null && !country.isBlank()) ? country : "India"
                );
                clients.add(clientRequest);
            }
        } catch (Exception e) {
            log.error("Failed to parse client Excel file", e);
            throw new RuntimeException("Failed to parse uploaded Excel file", e);
        }

        return clients;
    }

    private String getCellValueAsString(Cell cell) {
        if (cell == null) {
            return null;
        }
        String formatted = dataFormatter.formatCellValue(cell);
        return (formatted != null && !formatted.isBlank()) ? formatted.trim() : null;
    }

    private LocalDate parseDate(Cell cell) {
        if (cell == null) {
            return null;
        }
        if (cell.getCellType() == CellType.NUMERIC && DateUtil.isCellDateFormatted(cell)) {
            Date date = cell.getDateCellValue();
            if (date != null) {
                return date.toInstant().atZone(ZoneId.systemDefault()).toLocalDate();
            }
        }
        String str = getCellValueAsString(cell);
        if (str == null || str.isBlank()) {
            return null;
        }
        try {
            return LocalDate.parse(str, DateTimeFormatter.ISO_LOCAL_DATE);
        } catch (Exception e) {
            return null;
        }
    }

    private Gender parseGender(String genderStr) {
        if (genderStr == null || genderStr.isBlank()) {
            return null;
        }
        try {
            return Gender.valueOf(genderStr.trim().toUpperCase());
        } catch (Exception e) {
            return null;
        }
    }

    private boolean isRowEmpty(Row row) {
        if (row == null) {
            return true;
        }
        short firstCell = row.getFirstCellNum();
        short lastCell = row.getLastCellNum();
        if (firstCell < 0 || lastCell < 0) {
            return true;
        }
        for (int c = firstCell; c < lastCell; c++) {
            Cell cell = row.getCell(c);
            if (cell != null && cell.getCellType() != CellType.BLANK) {
                String val = getCellValueAsString(cell);
                if (val != null && !val.isBlank()) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: backend-nodejs/src/modules/customer/services/clientExcelService.ts
import ExcelJS from 'exceljs';
import { CreateClientDto } from '../dto/clientDto';
import { Gender } from '../enums/clientEnums';
import { logger } from '../../../common/utils/logger';

export class ClientExcelService {
  /**
   * Generates a styled Excel (.xlsx) template with headers and 1 sample row.
   * RMs download this template, fill in client records, and upload it.
   */
  async generateClientBulkTemplate(): Promise<Buffer> {
    const workbook = new ExcelJS.Workbook();
    const worksheet = workbook.addWorksheet('Clients Template');

    // Define column headers, keys, and display widths
    worksheet.columns = [
      { header: 'First Name', key: 'firstName', width: 18 },
      { header: 'Last Name', key: 'lastName', width: 18 },
      { header: 'Email', key: 'email', width: 30 },
      { header: 'Phone', key: 'phone', width: 18 },
      { header: 'PAN', key: 'pan', width: 16 },
      { header: 'Date of Birth (YYYY-MM-DD)', key: 'dateOfBirth', width: 26 },
      { header: 'Gender (MALE/FEMALE/OTHER)', key: 'gender', width: 28 },
      { header: 'Address Line', key: 'addressLine', width: 30 },
      { header: 'City', key: 'city', width: 18 },
      { header: 'State', key: 'state', width: 18 },
      { header: 'Pincode', key: 'pincode', width: 14 },
      { header: 'Country', key: 'country', width: 16 },
    ];

    // Header styling: Dark Blue corporate theme with bold white text
    const headerRow = worksheet.getRow(1);
    headerRow.font = { bold: true, color: { argb: 'FFFFFFFF' } };
    headerRow.fill = {
      type: 'pattern',
      pattern: 'solid',
      fgColor: { argb: 'FF1E3A8A' },
    };
    headerRow.alignment = { vertical: 'middle', horizontal: 'center' };
    headerRow.height = 24;

    // Provide 1 sample row to demonstrate expected formatting
    worksheet.addRow({
      firstName: 'Aarav',
      lastName: 'Sharma',
      email: 'aarav.sharma@example.com',
      phone: '9876543210',
      pan: 'ABCDE1234F',
      dateOfBirth: '1992-05-15',
      gender: 'MALE',
      addressLine: 'Flat 402, Highline Towers',
      city: 'Bengaluru',
      state: 'Karnataka',
      pincode: '560001',
      country: 'India',
    });

    const buffer = await workbook.xlsx.writeBuffer();
    return Buffer.from(buffer);
  }

  /**
   * Parses uploaded Excel buffer into validated CreateClientDto records.
   */
  async parseClientExcel(buffer: Buffer): Promise<CreateClientDto[]> {
    const workbook = new ExcelJS.Workbook();
    await workbook.xlsx.load(buffer as unknown as Parameters<typeof workbook.xlsx.load>[0]);

    const worksheet = workbook.worksheets[0];
    if (!worksheet) {
      logger.warn('Excel parsing aborted: No worksheet found');
      return [];
    }

    const clients: CreateClientDto[] = [];

    // Row 1 is header; iterate starting from row 2
    worksheet.eachRow((row, rowNumber) => {
      if (rowNumber === 1) return;

      const firstName = this.getCellValue(row.getCell(1));
      const email = this.getCellValue(row.getCell(3));
      const phone = this.getCellValue(row.getCell(4));

      // Skip empty or invalid rows missing primary required fields
      if (!firstName || !email || !phone) {
        return;
      }

      const lastName = this.getCellValue(row.getCell(2)) || '';
      const pan = this.getCellValue(row.getCell(5))?.toUpperCase();
      const rawDob = row.getCell(6).value;
      const dateOfBirth = this.parseDate(rawDob);
      const rawGender = this.getCellValue(row.getCell(7))?.toUpperCase();
      const gender = Object.values(Gender).includes(rawGender as Gender)
        ? (rawGender as Gender)
        : undefined;

      const addressLine = this.getCellValue(row.getCell(8));
      const city = this.getCellValue(row.getCell(9));
      const state = this.getCellValue(row.getCell(10));
      const pincode = this.getCellValue(row.getCell(11));
      const country = this.getCellValue(row.getCell(12)) || 'India';

      clients.push({
        firstName,
        lastName,
        email: email.toLowerCase(),
        phone,
        pan: pan || undefined,
        dateOfBirth,
        gender,
        addressLine: addressLine || undefined,
        city: city || undefined,
        state: state || undefined,
        pincode: pincode || undefined,
        country,
      });
    });

    logger.info({ parsedCount: clients.length }, 'Successfully parsed clients from Excel');
    return clients;
  }

  /**
   * Helper: Extracts string representation using ExcelJS cell.text.
   */
  private getCellValue(cell: ExcelJS.Cell): string | undefined {
    const text = cell.text?.trim();
    return text && text.length > 0 ? text : undefined;
  }

  /**
   * Helper: Safely parses Excel dates using unknown type and runtime type-guards.
   */
  private parseDate(val: unknown): Date | undefined {
    if (!val) return undefined;
    if (val instanceof Date) return val;
    if (typeof val === 'string' || typeof val === 'number') {
      const parsed = new Date(val);
      return isNaN(parsed.getTime()) ? undefined : parsed;
    }
    return undefined;
  }
}

export const clientExcelService = new ClientExcelService();
